core/sector_storage/stores/impl/index_impl.cpp

```diff
@@ -0,0 +1,305 @@
+      const SectorId &sector,
+      const fc::primitives::sector_file::SectorFileType &file_type,
+      bool allow_fetch) {
+    std::shared_lock lock(mutex_);
+    std::unordered_map<ID, uint64_t> storage_ids;
+
+    for (const auto &type : primitives::sector_file::kSectorFileTypes) {
+      if ((file_type & type) == 0) {
+        continue;
+      }
+
+      std::string sector_id = toSectorsID(sector, type);
+      auto sector_iter = sectors_.find(sector_id);
+      if (sector_iter == sectors_.end()) {
+        continue;
+      }
+      for (const auto &id : sector_iter->second) {
+        storage_ids[id]++;
+      }
+    }
+
+    std::vector<StorageInfo> result;
+    for (const auto &[id, count] : storage_ids) {
+      if (stores_.find(id) == stores_.end()) {
+        // logger
+        continue;
+      }
+
+      auto store = stores_[id].info;
+
+      for (uint64_t i = 0; i < store.urls.size(); i++) {
+        HttpUri uri;
+        try {
+          uri.parse(store.urls[i]);
+        } catch (const std::runtime_error &err) {
+          return IndexErrors::InvalidUrl;
+        }
+        boost::filesystem::path path = uri.path();
+        path = path / toString(file_type) / sectorName(sector);
+        uri.setPath(path.string());
+        store.urls[i] = uri.str();
+      }
+
+      store.weight = store.weight * count;
+      result.push_back(store);
+    }
+
+    if (allow_fetch) {
+      for (const auto &[id, storage_info] : stores_) {
+        if (storage_ids.find(id) != storage_ids.end()) {
+          continue;
+        }
+        auto store = stores_[id].info;
+
+        for (uint64_t i = 0; i < store.urls.size(); i++) {
+          HttpUri uri;
+          try {
+            uri.parse(store.urls[i]);
+          } catch (const std::runtime_error &err) {
+            return IndexErrors::InvalidUrl;
+          }
+          boost::filesystem::path path = uri.path();
+          path = path / toString(file_type) / sectorName(sector);
+          uri.setPath(path.string());
+          store.urls[i] = uri.str();
+        }
+
+        store.weight = 0;
+        result.push_back(store);
+      }
+    }
+
+    return result;
+  }
+
+  outcome::result<std::vector<StorageInfo>> SectorIndexImpl::storageBestAlloc(
+      const fc::primitives::sector_file::SectorFileType &allocate,
+      fc::primitives::sector::RegisteredProof seal_proof_type,
+      bool sealing) {
+    std::shared_lock lock(mutex_);
+
+    OUTCOME_TRY(
+        req_space,
+        fc::primitives::sector_file::sealSpaceUse(allocate, seal_proof_type));
+
+    std::vector<StorageEntry> candidates;
+
+    for (const auto &[id, storage] : stores_) {
+      if (sealing && !storage.info.can_seal) {
+        continue;
+      }
+      if (!sealing && !storage.info.can_store) {
+        continue;
+      }
+
+      if (req_space > storage.fs_stat.available) {
+        continue;
+      }
+
+      if (system_clock::now() - storage.last_heartbreak
```

Comment:
<img width="76" alt="image" src="https://user-images.githubusercontent.com/36847043/82985347-8f1f7580-9ffc-11ea-96a7-fc9a0e74e384.png">

should it be a heartbeat?

Reply:
Fixed

Reply:
nope?